src/data-structures/heap/MaxHeapAdhoc.js

```diff
@@ -16,15 +17,17 @@ class MaxHeapAdhoc {
   }
 
   peek() {
-    return this.heap[0];
+    return this.isEmpty() ? undefined : this.heap[0];
   }
 
   poll() {
-    if (this.heap.length === 0) return undefined;
+    if (this.isEmpty()) return undefined;
     const top = this.heap[0];
-    this.heap[0] = this.heap[this.heap.length - 1];
-    this.heap.pop();
-    this.heapifyDown();
+    const last = this.heap.pop(); // Remove the last element to replace the top
+    if (!this.isEmpty()) {
+      this.heap[0] = last; // Move the last element to the root
+      this.heapifyDown(); // Re-establish the heap property
```

Comment:
Why do we explicitly check? `!this.isEmpty()` here at line no 27 as we are returning undefined if `this.isEmpty()` is true.
So the lines 25 to 29 will execute only if `!this.isEmpty()` is true.  This check is redundant here.